iwyu_ast_util.cc

```diff
@@ -546,32 +558,32 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
+    TraverseTypeHelper(type);
     return Base::TraverseType(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
+    TraverseTypeHelper(type_loc.getType());
     return Base::TraverseTypeLoc(type_loc);
   }
 
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
-  }
-
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
+  void TraverseTypeHelper(QualType qual_type) {
+    if (qual_type.isNull())
       return;
 
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+    const Type* const type = qual_type.getTypePtr();
+    const Type* const desugared_until_alias = Desugar(type);
+    seen_types_.insert(desugared_until_alias);
+
+    const Type* const desugared =
+        Desugar(GetNonAliasType(desugared_until_alias));
```

Comment:
Also a bit curious about this `Desugar` -- once the alias types are gone, do you expect any other sugar to remain? Could be `ElaboratedType` I suppose...

Reply:
And doing a full desugar here using `Type::getUnqualifiedDesugaredType` would strip too much (`TemplateSpecializationType`)?

Reply:
> Also a bit curious about this Desugar -- once the alias types are gone, do you expect any other sugar to remain?

Yes, there may still be a lot of sugar! E.g., for such a code:
```cpp
namespace ns {
  using ::Indirect;
}
ns::Indirect i;
typedef decltype(i) Typedef;
```
clang constructs such a tree:
```
TypedefType 0x18e19b10400 'Typedef' sugar
|-Typedef 0x18e19b10260 'Typedef'
`-DecltypeType 0x18e19b10220 'decltype(i)' sugar
  |-DeclRefExpr 0x18e19b101f8 'ns::Indirect':'struct Indirect' lvalue Var 0x18e19b0c688 'i' 'ns::Indirect':'struct Indirect' non_odr_use_unevaluated
  `-ElaboratedType 0x18e19b0c610 'ns::Indirect' sugar
    `-UsingType 0x18e19b0c5e0 'Indirect' sugar
      |-UsingShadow 0x18e19b0c538 'Indirect'
      `-RecordType 0x18e19afbf10 'struct Indirect'
        `-CXXRecord 0x18e19afbe60 'Indirect'
```
> And doing a full desugar here using Type::getUnqualifiedDesugaredType would strip too much (TemplateSpecializationType)?

Yes, I think so...

Reply:
> Also a bit curious about this Desugar -- once the alias types are gone, do you expect any other sugar to remain?

Of course, this `Desugar` may be moved into `GetNonAliasType`, but I think it would be excessive responsibility for `GetNonAliasType` (and three places of `Desugar` call instead of one).

Btw, `desugared_until_alias` should better be renamed into something like `desugared_until_alias_or_tpl_spec`.

Reply:
I've realized that this version of `GetNonAliasType`, of course, doesn't work as expected for recursive typedefs, like these:
```cpp
typedef Indirect Typedef1;
typedef Typedef1 Typedef2;
```
at least due to `ElaboratedType` between them in the tree. `Desugar` should really be placed into `GetNonAliasType`.